### ISSUE TITLE
`.max(Tensor([ 1, nan]))` should be nan 

### ISSUE THREAD

**Issue by xl0** [geohot/tinygrad#862]
`Tensor([1, np.nan]).numpy().max()` -> `np.nan`
`Tensor([1, np.nan]).max().numpy()` -> `array([1.], dtype=float32)`

Wrong output shape aside, it should be `nan`, not `1`.

**Comment by wozeparrot** [comment]
I just tested this on all the backends I can test on, and it only has the wrong output on the GPU backend.

CLANG uses `((x>y)?x:y)` as its max.
GPU uses opencl `max` which is undefined when arguments are nan.

If I use CLANG's max on GPU it works correctly.

**Comment by geohot** [comment]
Nope, I think we can change it on GPU + add a test.

**Comment by flammit** [comment]
I think the deeper question is what tinygrad will accept as the desired result for BinaryOps.MAX or ReduceOps.MAX.

In the presence of some NaN but not all NaNs, I believe the following MAX behaviors can currently be supported:

| BACKEND | MAX is not NaN | MAX is NaN |
| --- | --- | --- |
| CPU | Yes (using np.nanmax) | Yes (using np.max/amax) |
| GPU | Yes (using fmax) | Not easily (vectorized select and isnans) |
| METAL | Yes (using max/fmax) | Not easily (vectorized select and isnans) |
| CLANG | Yes | Yes (scalar isnans) |
| LLVM | Yes | No (don't see how llvmlite can generate this) |
| TORCH | Not easily (missing torch.nanmax) | Yes (using torch.amax) | 

There's an issue tracking the lack of torch.nanmax here: https://github.com/pytorch/pytorch/issues/61474, even though there is a torch.fmax. The llvmlite IRBuilder doesn't have NaN detection, but llvm.maxnum matches the return non-NaN operand.

Here's some refs:
CPU: https://numpy.org/doc/stable/reference/generated/numpy.nanmax.html, https://numpy.org/doc/stable/reference/generated/numpy.max.html
GPU: https://man.opencl.org/commonMax.html, https://man.opencl.org/fmax.html
METAL: https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf page 161
CLANG: https://clang.llvm.org/docs/LanguageExtensions.html
LLVM: https://llvm.org/docs/LangRef.html#llvm-maxnum-intrinsic
TORCH: https://pytorch.org/docs/stable/generated/torch.fmax.html

Right now there's no choice that will pass on all backends. @geohot what do you think?

**Comment by karan-amuse** [comment]
I would assume we're trying to reproduce the same functionality of `numpy.max()`? Also it makes sense to propage NaNs for all associative operations to signal invalid results.

**Comment by karan0handa** [comment]
I would assume we're trying to reproduce the same functionality of numpy.max()? Also it makes sense to propage NaNs for all associative operations to signal invalid results.

**Comment by uncommonSensor** [comment]
It should be noted for WebGPU as well: "Parsing error: no definition in scope for identifier: 'nanf'".

Also, the initial commit for WebGPU skips a couple tests because "WebGPU does not have support for inf/nan"

**Comment by uncommonSensor** [comment]
More info:

I ran some tests on the different variations of 'max' in pytorch, tinygrad, and numpy:

Notes:

tinygrad max doesn't handle nans from what I can tell. So it's currently most like torch.fmax, except torch.fmax takes two tensors and gets max element-wise (ignoring nans).
np.nanmax() uses reduce but only returns nan if all elements in a tensor are nan, otherwise it ignores nan elements.
The original ticket was written up because tinygrad max is not handling nans, like numpy max does:

Tensor([1, np.nan]).numpy().max() -> np.nan
Tensor([1, np.nan]).max().numpy() -> array([1.], dtype=float32)

Wrong output shape aside, it should be nan, not 1.
It seems like the best option is either to add a nan check to the existing mlops Max class, or create a new max function. Calling it fmax could be confusing though.

Pytorch:

t1 = torch.tensor([2, 1, np.nan, np.nan])
t2 = torch.tensor([np.nan, 5, 3, np.nan])

torch.max(t1).numpy()               -> array(nan, dtype=float32)
torch.max(t1, t2).numpy()           -> array([nan,  5., nan, nan], dtype=float32)
torch.maximum(t1, t2).numpy()       -> array([nan,  5., nan, nan], dtype=float32)
torch.fmax(t1, t2).numpy()          -> array([ 2.,  5.,  3., nan], dtype=float32)
np.nanmax(t1)                       -> 2.0
Tinygrad:

t1 = Tensor([2, 1, np.nan, np.nan])
t2 = Tensor([np.nan, 5, 3, np.nan])

Tensor.max(t1).numpy()              -> array(2., dtype=float32)
Tensor.max(t1, t2).numpy()          -> array(2., dtype=float32)  **(This is comparable torch.fmax)**
Tensor.maximum(t1, t2).numpy()      -> array([ 2.,  5.,  3., nan], dtype=float32)
Numpy:

t1 =np.array([[np.nan, 5, 3, np.nan], [2, 2, 2, np.nan]])
np.nanmax(t1) -> 5.0

t1 = np.array([np.nan, np.nan, np.nan, np.nan]) 
np.nanmax(t1) -> nan

t1 = torch.tensor([1, np.nan, 5, 3])
np.nanmax(t1) -> 5.0

t1 = torch.tensor([np.nan, np.nan, np.nan, np.nan])
np.nanmax(t1) ->  <stdin>:1: RuntimeWarning: All-NaN axis encountered -> **nan**

**Comment by uncommonSensor** [comment]
**More info:**

 I ran some tests on the different variations of 'max' in pytorch, tinygrad, and numpy:

**Notes:**
- tinygrad max doesn't handle nans from what I can tell. So it's currently most like torch.fmax, except torch.fmax takes two tensors and gets max element-wise (ignoring nans). 
- np.nanmax() uses reduce but only returns nan if all elements in a tensor are nan, otherwise it ignores nan elements.

The original ticket was written up because tinygrad max is not handling nans, like numpy max does:
```
Tensor([1, np.nan]).numpy().max() -> np.nan
Tensor([1, np.nan]).max().numpy() -> array([1.], dtype=float32)

Wrong output shape aside, it should be nan, not 1.
```

It seems like the best option is either to add a nan check to the existing mlops Max class, or create a new max function. Calling it fmax could be confusing though.

**Pytorch:**
```
t1 = torch.tensor([2, 1, np.nan, np.nan])
t2 = torch.tensor([np.nan, 5, 3, np.nan])

torch.max(t1).numpy()               -> array(nan, dtype=float32)
torch.max(t1, t2).numpy()           -> array([nan,  5., nan, nan], dtype=float32)
torch.maximum(t1, t2).numpy()       -> array([nan,  5., nan, nan], dtype=float32)
torch.fmax(t1, t2).numpy()          -> array([ 2.,  5.,  3., nan], dtype=float32)
np.nanmax(t1)                       -> 2.0
```

**Tinygrad:**
```
t1 = Tensor([2, 1, np.nan, np.nan])
t2 = Tensor([np.nan, 5, 3, np.nan])

Tensor.max(t1).numpy()              -> array(2., dtype=float32)
Tensor.max(t1, t2).numpy()          -> array(2., dtype=float32)  **(This is comparable torch.fmax)**
Tensor.maximum(t1, t2).numpy()      -> array([ 2.,  5.,  3., nan], dtype=float32)
```

**Numpy:**
```
t1 =np.array([[np.nan, 5, 3, np.nan], [2, 2, 2, np.nan]])
np.nanmax(t1) -> 5.0

t1 = np.array([np.nan, np.nan, np.nan, np.nan]) 
np.nanmax(t1) -> nan

t1 = torch.tensor([1, np.nan, 5, 3])
np.nanmax(t1) -> 5.0

t1 = torch.tensor([np.nan, np.nan, np.nan, np.nan])
np.nanmax(t1) ->  <stdin>:1: RuntimeWarning: All-NaN axis encountered -> **nan**

```

**Comment by JasonOA888** [comment]
I analyzed this issue. The nan handling in `max()` reduction is a known limitation.

**Current state:**
- Test marked as `@unittest.expectedFailure` in `test/backend/test_edgecases.py:39`
- Test skipped in `test/backend/test_ops.py:2861`

**Root cause:**
The `Ops.MAX` reduction in GPU kernels does not implement IEEE 754 compliant nan propagation. The AMD pcode simulator (`test/mockgpu/amd/pcode.py:124`) shows the correct approach:

```python
if dt == dtypes.float32: 
    result = _isnan(result).where(b, _isnan(b).where(result, minmax(result, b)))
```

But this nan-aware logic is not applied in actual GPU code generation for reduce operations.

**Fix complexity:**
This requires changes in:
1. GPU kernel generation (codegen/)
2. Backend-specific implementations (CUDA, HIP, Metal, etc.)
3. Potentially affects performance for all max/min operations

This is a non-trivial fix that trades correctness vs performance for floating-point reductions.

**Comment by Yaskophx** [comment]
Should `Tensor.max()` propagate NaN like `torch.max/np.max`, or ignore it like `torch.fmax/np.fmax`?

The [MAX decomposition](https://github.com/tinygrad/tinygrad/blob/master/tinygrad/uop/decompositions.py#L448) drops NaN because `(a < NaN)` is always False, but the [AMD pcode](https://github.com/tinygrad/tinygrad/blob/master/test/mockgpu/amd/pcode.py#L124) explicitly ignores NaN too.
